sprinklers/base.py

```diff
@@ -44,23 +59,70 @@ def split_queryset(self, queryset, chunk_size=chord_size):
             else:
                 chunk.append(item)
 
-        yield chunk
+    def shard_start(self, shard_id, from_pk=None, to_pk=None):
+        pks = self.get_queryset_pks(from_pk, to_pk)
 
-    def start(self):
-        qs = self.get_queryset()
+        c = chord(
+            (
+                _async_subtask.s(pk, self.__class__.__name__, self.kwargs).set(queue=self.get_subtask_queue()).set(queue=self.get_subtask_queue())
+                for pk in pks
+            ),
+            _sprinkler_shard_finished_wrap.s(sprinkler_name=self.__class__.__name__, shard_id=shard_id, kwargs=self.kwargs).set(queue=self.get_subtask_queue())
+        )
 
-        if self.chord_size is not None:
-            chunks = self.split_queryset(qs, self.chord_size)
-        else:
-            chunks = [qs]
+        start_time = time()
+        c.apply_async()
+        end_time = time()
+
+        duration = (end_time - start_time) * 1000
+        self.log(f"Started shard {shard_id} in {duration}ms.")
+
+        return shard_id
+
+    def shard_finished(self, shard_id, results):
+        pass
 
-        for chunk in chunks:
-            ids = [o['id'] if isinstance(o, dict) else o.id for o in chunk]
+    def get_queryset_pks(self, from_pk=None, to_pk=None):
+        queryset = self.get_queryset().only('pk').order_by('pk')
+
+        if from_pk is not None:
+            queryset = queryset.filter(pk__gt=from_pk)
+
+        if to_pk is not None:
+            queryset = queryset.filter(pk__lte=to_pk)
+
+        # values_list in django 1.11 is broken and will run out of memory when iterating over a large queryset, even with .iterator()
+        # the following code does basically the same thing as values_list, without running out of memory
+        db = queryset.db
+        compiler = queryset.query.get_compiler(db)
+        results = compiler.execute_sql(chunked_fetch=True)
+
+        for row in compiler.results_iter(results):
```

Comment:
Nice! 
Can we use `ModelIterable` from [here](https://docs.djangoproject.com/en/1.11/_modules/django/db/models/query/)  ? 
And wrap queryset with it ?

Reply:
I'd rather not--the problem is the model iterable returns a model and I don't want to instantiate a model just to shard, I only want the PK

Reply:
FYI instantiating the model makes the sharding take about 4 times as long, even if it's with only the PK.